lib/middleware.py

```diff
@@ -1,19 +1,27 @@
 from django.conf import settings
-from django.http import HttpResponseRedirect
+from django.http import Http404, HttpResponseRedirect
 from django.utils import translation
 from django.utils.cache import patch_vary_headers
 from django.utils.deprecation import MiddlewareMixin
+from django.shortcuts import redirect
+
+from urllib.parse import urlparse
 
 
 class LocaleMiddleware(MiddlewareMixin):
 
     response_redirect_class = HttpResponseRedirect
 
     def process_request(self, request):
-        if hasattr(request, 'user') and request.user.is_authenticated:
+        language = translation.get_language_from_request(request)
+        query_language = request.GET.get('hl')
+        if query_language:
+            language = query_language
+            lang_codes = [lang[0][:2] for lang in settings.LANGUAGES]
+            if language and (language[:2] not in lang_codes):
+                raise Http404('Page is not available in this language.')   
+        elif hasattr(request, 'user') and request.user.is_authenticated:
             language = request.user.userprofile.language
-        else:
-            language = translation.get_language_from_request(request)
```

Comment:
This `else` branch could be restored as it would make more sense with the current code. Would remove the execution of `get_language_from_request` in case it comes from somewhere else.